implementations/elixir/ockam/ockam/lib/ockam/secure_channel/key_establishment_protocol/xx/protocol.ex

```diff
@@ -119,55 +145,22 @@ defmodule Ockam.SecureChannel.KeyEstablishmentProtocol.XX.Protocol do
   end
 
   defp setup_message_payloads(options, state) do
-    state =
-      state
-      |> Map.put(:m1_payload, Keyword.get(options, :message1_payload, @default_m1_payload))
-      |> Map.put(:m2_payload, Keyword.get(options, :message2_payload, @default_m2_payload))
-      |> Map.put(:m3_payload, Keyword.get(options, :message3_payload, @default_m3_payload))
-
+    state = Map.put(state, :payloads, Keyword.get(options, :payloads, @default_payloads))
     {:ok, state}
   end
 
-  def encode(message_name, %{xx_key_establishment_state: protocol_state} = data)
-      when message_name in [:message1, :message2, :message3] do
-    encoder = String.to_existing_atom("encode_" <> Atom.to_string(message_name))
-    return_value = apply(__MODULE__, encoder, [protocol_state])
-
-    case return_value do
-      {:ok, encoded, protocol_state} ->
-        ## TODO: optimise double encoding of binaries
-        encoded_payload = :bare.encode(encoded, :data)
-        {:ok, encoded_payload, %{data | xx_key_establishment_state: protocol_state}}
-
-      {:error, reason} ->
-        {:error, {:failed_to_encode, message_name, reason}}
-    end
-  end
-
-  def decode(message_name, message_payload, %{xx_key_establishment_state: protocol_state} = data)
-      when message_name in [:message1, :message2, :message3] do
-    ## TODO: optimise double encoding of binaries
-    {:ok, message, ""} = :bare.decode(message_payload, :data)
-    decoder = String.to_existing_atom("decode_" <> Atom.to_string(message_name))
-    return_value = apply(__MODULE__, decoder, [message, protocol_state])
+  def encode(:message1, %{e: e, payloads: payloads} = state) do
+    payload = Map.get(payloads, :message1, "")
```

Comment:
Will we eventually want to remove the default value here after the three-message work, or is always reasonable to have a fallback to empty string?

Reply:
good question.   Even after the upcoming changes,  the _first_ handshake message won't carry any payload in our protocol.  So this could had even be hardwired to be the an empty binary. 
_But_   there are existing test vectors we use for unit tests of the noise handshake that _do_ have payloads..  so that's right now the only value of keeping it open (that we can validate existing test vectors for noise,  not only ones explicitly written by us)